### PR TITLE
Students can only review their own quiz attempt results

### PR DESCRIPTION
**Pull request by thomasplevy**
## Description

Fixes https://github.com/gocodebox/private-issues/issues/19

## How has this been tested?

+ Manually
+ New & existing phpunit tests

## Types of changes
+ Bugfix

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: tests/phpunit/unit-tests/user/class-llms-test-student-quizzes.php [hunk 1]
```diff
@@ -1,38 +1,64 @@
...
+	private function get_student_with_quizzes( $attempts = 3 ) {
 
 		$uid = $this->factory->user->create();
 		$student = llms_get_student( $uid );
-		$courses = $this->generate_mock_courses( 3, 1, 1, 1 );
+		$courses = $this->generate_mock_courses( $attempts, 1, 1, 1 );
 		$this->complete_courses_for_student( $uid, $courses );
 		return $student;
 
 	}
```
Comment (eri-trabiccolo) [inline tests/phpunit/unit-tests/user/class-llms-test-student-quizzes.php:59]: _oh gimme gimme some docs_

File: includes/models/model.llms.student.quizzes.php [hunk 2]
```diff
@@ -110,22 +120,39 @@ public function get_attempts_by_quiz( $quiz_id, $args = array() ) {
 	/**
 	 * Retrieve an attempt by attempt id
 	 *
-	 * @param    int $attempt_id  Attempt ID
-	 * @return   LLMS_Quiz_Attempt
-	 * @since    3.16.0
-	 * @version  3.16.0
+	 * @since 3.16.0
+	 * @since [version] Only return
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.student.quizzes.php]: typical case of _documentation interruptus_

File: includes/models/model.llms.student.quizzes.php [hunk 3]
```diff
@@ -45,47 +44,58 @@ public function count_attempts_by_quiz( $quiz_id ) {
...
 	 * Retrieve quiz attempts
 	 *
-	 * @param    int   $quiz_id  WP Post ID of the quiz
-	 * @param    array $args     additional args to pass to LLMS_Query_Quiz_Attempt
-	 * @return   LLMS_Quiz_Attempt[]
 	 * @since    3.16.0
-	 * @version  3.16.0
+	 *
+	 * @param int   $quiz_id  WP Post ID of the quiz.
+	 * @param array $args     Additional args to pass to LLMS_Query_Quiz_Attempt.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.student.quizzes.php]:
```suggestion
	 * @param int   $quiz_id WP Post ID of the quiz.
	 * @param array $args    Additional args to pass to LLMS_Query_Quiz_Attempt.
```

File: includes/models/model.llms.student.quizzes.php [hunk 4]
```diff
@@ -110,22 +120,39 @@ public function get_attempts_by_quiz( $quiz_id, $args = array() ) {
...
 	 * Decodes an attempt string and returns the associated attempt
 	 *
-	 * @param    string $attempt_key  encoded attempt key
-	 * @return   LLMS_Quiz_Attempt|false
-	 * @since    3.9.0
-	 * @version  3.16.0
+	 * @since 3.9.0
+	 * @since 3.16.0 Unknown.
+	 *
+	 * @param string $attempt_key  Encoded attempt key.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.student.quizzes.php]:
```suggestion
	 * @param string $attempt_key Encoded attempt key.
```

File: tests/phpunit/unit-tests/user/class-llms-test-student-quizzes.php [hunk 5]
```diff
@@ -1,38 +1,64 @@
...
- * @group    quizzes
- * @group    LLMS_Student
- * @since    3.9.0
- * @version  3.16.11
+ *
+ * @group quizzes
+ * @group student_quizzes
+ * @group LLMS_Student
+ *
+ * @since 3.9.0
```
Comment (eri-trabiccolo) [inline tests/phpunit/unit-tests/user/class-llms-test-student-quizzes.php:9]: missing version tag?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
changes to doc stuff requested.

This PR fixes the issue it aims to fix (similar thing should be done for assignments) but introduces a regression for the groups add-on.
Because it allowed to view other members attempts, although it was not ideal because any user could, while we should allow that only to users who can.

The thing is that we should be able to initialize the `LLMS_Student_Quizzes` with the user id we need, and then check on whether the current user can view others (students|reports)?

What do you think?
(I'm sorry this is not the easy win you were looking for :P)

**Comment by thomasplevy** [comment]
@eri-trabiccolo thanks for catching this. I've reworked this to introduce a new capability, `view_grades` which is used by the core to allow admins & managers to view anyone's quiz results (which they can already do on the frontend). Instructors will be able to view results for students on their content.

The core will not implement any true frontend for these users to arrive on these screens but it's implemented in a "forward compatible" way as at some point I can see us introducing reporting for instructors/others like the groups plugin currently does.

Users can always view their own results.

The groups plugin is adding it's own handling via https://github.com/gocodebox/lifterlms-groups/pull/141 to allow group members who can view group reports to continue viewing quiz results as expected

File: includes/class.llms.user.permissions.php [hunk 6]
```diff
@@ -159,6 +159,57 @@ public static function get_editable_roles() {
...
 	}
 
+	/**
+	 * Modify a users ability to `view_grades`
+	 *
+	 * Users can view the grades (quiz results) if one of the following conditions is met:
+	 *   + Users can view their own grades.
+	 *   + Admins and LMS Managers can view anyone's grade.
+	 *   + Any user who has been explicitly granted the `view_grades` cap can view anyone's grade (via custom code).
+	 *   + Any user who can `edit_post` for the course the quiz belongs to can view grades of the students within that course.
```
Comment (eri-trabiccolo) [inline includes/class.llms.user.permissions.php]: according to the code... this is true assuming the user is an instructor or an instructor assistant, right?

**Review by eri-trabiccolo — APPROVED** [review]
Looks fine to me.
I left a comment (question) though.

File: includes/class.llms.user.permissions.php (hunk 6 above)
Comment (thomasplevy) [reply includes/class.llms.user.permissions.php]: Yes, updating the comment thanks

Comment (thomasplevy) [inline includes/class.llms.user.permissions.php]:
```suggestion
	 *   + Any instructor/assistant who can `edit_post` for the course the quiz belongs to can view grades of the students within that course.
```

**Pull request merged**